### PR TITLE
Quote template parameters in cloudformation deploy

### PR DESCRIPTION
**Pull request by abrooksv**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Related Issue(s)
#2007 

## Testing
```
Deploying with following values
	===============================
	Stack name                 : HHHH
	Region                     : us-west-2
	Confirm changeset          : False
	Deployment s3 bucket       : fooooooo-test-us-west-2
	Capabilities               : ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
	Parameter overrides        : {'pCron': 'cron(30 0 * * ? *)'}
```

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt [hunk 1]
```diff
@@ -96,19 +99,12 @@ class SamDeployDialog(
...
-                .withParameters("--build-dir")
-                .withParameters(buildDir.toString())
-                .apply {
-                    if (useContainer) {
-                        withParameters("--use-container")
-                    }
-                }
-
-            it
+            it.samBuildCommand(
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt:102]: I like this much better

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt [hunk 2]
```diff
@@ -209,20 +194,10 @@ class SamDeployDialog(
...
-                is ExecutableInstance.Executable -> it
-                else -> throw RuntimeException((it as? ExecutableInstance.BadExecutable)?.validationError)
-            }
-            return@thenApply samExecutable
-                .getCommandLine()
-                .withWorkDirectory(template.parent.path)
-                .withEnvironment(envVars)
+    private fun createBaseCommand(): CompletionStage<GeneralCommandLine> = ExecutableManager.getInstance().getExecutable<SamExecutable>().thenApply {
+        when (it) {
+            is ExecutableInstance.Executable -> return@thenApply it.getCommandLine()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt:199]: do you even need the `return` here? I'm assuming it should use the last evaluated expression.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt [hunk 3]
```diff
@@ -56,3 +59,74 @@ class SamExecutable : ExecutableType<SemVer>, AutoResolvable, Validatable {
...
+        addParameter("--capabilities")
+        addParameters(capabilities.map { it.capability })
+    }
+
+    addParameter("--no-execute-changeset")
+
+    if (parameters.isNotEmpty()) {
+        addParameter("--parameter-overrides")
+        parameters.forEach { (key, value) ->
+            addParameter("\"$key\"=\"$value\"")
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: Some additional things I'm concerned about:
What if the value already has a `"`, what if it already has an escaped quote `\"`, what if the value is already quoted, what if the value is already single quoted, what if the value has a single quote in it

Comment (justinmk3) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: Is there validation before this point? I would suggest throwing an error if the key or value contains any of `\`, `"`, or `=` (and probably more).

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt (hunk 2 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/SamDeployDialog.kt:199]: Cleaned it up, its gone now

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: I am not sure about '=', but \" is supported by SAM

https://github.com/wchengru/aws-sam-cli/blob/develop/tests/unit/cli/test_types.py#L65

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: Okay...I have come to the conclusion that getting this right is impossible due to bugs in SAM:

Expected: `Hello\ World`
| Tried                              | Result        |
|------------------------------------|---------------|
| `"Hello\ World"`    | `Hello World`   |
| `"Hello\\ World"`  | `Hello World`   |
| `"Hello\\\ World"` | `Hello\\ World` |

And `Hello\World` will always become `Hello\\World`

I have also have had to switch to single quotes if already quoted due a bug where they will strip off all surrounding quotes, not just the outside ones. (This requires SAM CLI 1.1.0...)

SAM is happy with `=` it seems:
`ParameterValue=\"Hello=World\" -> `'Hello=World'`

Comment (hunterwerlla) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: Can you make the box validate that? Quoting is way more complicated than it has any right to be

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutable.kt:129]: I am not sure we should, as far as a I can tell it is perfectly valid in CFN to do it.

I got the quotes issues all fixed but backslashes are broken now. I plan on cutting an issue to them and seeing what they have to say.

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutableTest.kt [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                "--template",
+                "$templatePath",
+                "--build-dir",
+                "$buildDir"
+            ).joinToString(separator = " ")
+        )
+
+        assertThat(cmd.workDirectory).isEqualTo(tempFolder.root)
+
+        assertThat(cmd.environment).containsEntry("Foo", "Bar")
```
Comment (hunterwerlla) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutableTest.kt:42]: Should also check the common environment variables for the credentials

**Review by hunterwerlla — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutableTest.kt (hunk 4 above)
Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/lambda/sam/SamExecutableTest.kt:42]: We dont pass creds at this layer, its already in the GeneralCommandLine before this is invoked.

**Pull request merged**